### ISSUE TITLE
DataSets is hard to use with Distributed.jl

### ISSUE THREAD

**Issue by c42f** [JuliaComputing/DataSets.jl#14]
A long distributed job is likely to want to open a dataset on each node and keep it open for the whole computation.

The use pattern in Distributed seems to be many RPCs driven by the main node, with the main node kicking off further computation as required, and with Distributed itself owning the event loop which drives the communication.

However, DataSets package mostly supports the scoped form

```julia
open(T, dataset("name") do data
    work(data)
end
```

The idea was that this would wrap around the program at a high level, but this seems largely inconsistent with the fact that one needs to exit the scope to return results to the Distributed event loop. #12 should help with this, by making unscoped `open()` more reliable and providing another way to manage resources.

A separate usability problem is that it's impossible to serialize handles like `Blob` or `BlobTree` when they rely on the availability of local resources such as disk caches which are only available on the main node. One is then forced into less natural use patterns such as sending paths between the nodes.

**Comment by c42f** [comment]
To quickly document the current workaround to #8... it's not pretty:

```julia
ds = dataset("some_name")
ch = Channel()
@async try
    open(BlobTree, ds) do t
        put!(ch, t)
        take!(ch)
    end
catch exc
    @error "Error opening dataset" exception=(exc,catch_backtrace())
    close(ch)
end
tree = take!(ch)
```

**Comment by c42f** [comment]
This is mostly solved by #12, I've opened #24 to track the remaining parts.
